### PR TITLE
ingress: Fix conformance tests for host-rules and path-rule

### PR DESCRIPTION
**Pull request by sayboras**
### Description

Please refer to individual commit for more details.

I am very much on the fence for the first commit. Basically, Ingress conformance test didn't really enforce TLS only, hence there will be 2 listeners (e.g. 80 and 443), possible options can be considered:
- keep the behavior aligned with k8s Ingress [conformance test](https://github.com/kubernetes-sigs/ingress-controller-conformance)
- always enforce HTTPs and return 308 http code with https location. One drawback is that conformance test will fail for 3 related test.
- add flag to control behavior (e.g. --enforce-ingress-https) :heavy_check_mark: 
   - default value can be true. This is the approach that nginx controller is taking. I have a local branch for this approach, so can easily swap it out. :heavy_check_mark: 
   - default value is set as false, let user set it accordingly.
 
 ### Testing
 
Testing was done locally with minikube, all the tests are passing now.

### CODE REVIEW

**Review by michi-covalent — APPROVED** [review]
i vote for

> keep the behavior aligned with k8s Ingress [conformance test](https://github.com/kubernetes-sigs/ingress-controller-conformance)

edit:

i'm also fine with this option:

> add flag to control behavior (e.g. --ingress-enforce-https)

i think the default should be false so that the conformance tests pass by default.

File: operator/pkg/ingress/envoy_config.go
Comment (nebril) [inline operator/pkg/ingress/envoy_config.go]: Will this behaviour be correct if `ingressPath.Path` is equal to `"/"`?

File: operator/pkg/ingress/envoy_config.go
Comment (nebril) [inline operator/pkg/ingress/envoy_config.go]: Should we also sort exact matches by length, or does it not matter?

e.g. should `/foobar` be before `/foo`?

File: operator/pkg/ingress/envoy_config.go
Comment (nebril) [inline operator/pkg/ingress/envoy_config.go]: nit: if `rule.HTTP.Paths` length is known, you can initialize `routes` with `make(SortableRoute,0 , len(rule.HTTP.Paths))` to avoid superfluous memory allocations while appending on line 436

**Review by nebril — COMMENTED** [review]
I am not super comfortable reviewing Envoy-related bits, but most of the PR looks good to me. Left some comments and questions inline.

File: operator/pkg/ingress/envoy_config.go
Comment (sayboras) [reply operator/pkg/ingress/envoy_config.go]: good point :100:

File: operator/pkg/ingress/envoy_config.go
Comment (sayboras) [reply operator/pkg/ingress/envoy_config.go]: the exactmatch order didn't matter, as envoy will only come to one match consistently :)

File: operator/pkg/ingress/envoy_config.go
Comment (sayboras) [reply operator/pkg/ingress/envoy_config.go]: yup, if the ingress is having prefix as "/", ingress will basically match every paths e.g. http://foo.com/, http://foo.com/, or http://foo.com/abc

In this case, envoy prefix route will be (/.*)?$ which is aligned with Ingress.

**Comment by sayboras** [comment]
All tests are passed, small force pushed to address review comment with below difft

```
operator/pkg/ingress/envoy_config.go --- Go
453 }                                                                                    453 }                                                                                    
454                                                                                      454                                                                                      
455 func getVirtualHost(ingress *slim_networkingv1.Ingress, rule slim_networkingv1.Ingre 455 func getVirtualHost(ingress *slim_networkingv1.Ingress, rule slim_networkingv1.Ingres
... ssRule) *envoy_config_route_v3.VirtualHost {                                         ... sRule) *envoy_config_route_v3.VirtualHost {                                          
456     var routes SortableRoute                                                         456     routes := make(SortableRoute, 0, len(rule.HTTP.Paths))                           
457     for _, path := range rule.HTTP.Paths {                                           457     for _, path := range rule.HTTP.Paths {                                           
458         route := envoy_config_route_v3.Route{                                        458         route := envoy_config_route_v3.Route{                                        
459             Match: getRouteMatch(rule.Host, path),                                   459             Match: getRouteMatch(rule.Host, path),                                   
```

File: operator/pkg/ingress/envoy_config.go
Comment (jrajahalme) [inline operator/pkg/ingress/envoy_config.go]: You can't use characters that would be interpreted as regex non-literal matches (such as `.` (dot)) into a regex, as it will match any character. For example, the above will translate `*.google.com` to  regex `[^.]+.google.com` which will match `notgoogle.com`!

So you'll have to:
- replace all dots with `[.]` to enforce matching only a dot
- make sure there are no wildcards (`*`) anywhere else than in the beginning, and that it is followed by a `.`
- generally anchoring the regex is a good idea, unless you find documentation that Envoy will not match a substring (meaning that Envoy anchors the regex for you)
- Bonus point for adding a (unit) test case verifying that above, as it can be a security hazard if this is not done right.

**Review by jrajahalme — CHANGES_REQUESTED** [review]
I'd opt for the more secure stance with `--ingress-enforce-https=true` by default, and then we can run the conformance test with `--ingress-enforce-https=false`.

Also the translation of the host wildcard to a regex needs some attention.

**Comment by sayboras** [comment]
Convert to draft to address the above two comments, thanks for your review :dancers:

File: operator/pkg/ingress/envoy_config.go
Comment (jrajahalme) [reply operator/pkg/ingress/envoy_config.go]: So the input string must be validated to only contain valid characters for a domain name (https://www.ietf.org/rfc/rfc1034.txt):
- [a-zA-Z0-9-]
- We've also had a case where users were using underscores (`_`) in domain names, so allow also those.
- replace potential `*.` in the beginning with `[^.]+[.]`
- replace each remaining `.` with `[.]`

Comment (sayboras) [reply operator/pkg/ingress/envoy_config.go]: do we need to do the first and second validation, as host in ingress spec is following https://datatracker.ietf.org/doc/html/rfc3986, which also conforms with 1034 ?

**Comment by sayboras** [comment]
Compared to the first version, the below points are done:
- add flag for enforce ingress https, changes are done in helm to make ingress controller config more flexible
- add unit tests for each changes
- incorporate the review comments from Jarno as per above.

Conformance tests are all passing still :)

File: install/kubernetes/cilium/values.yaml
Comment (qmonnet) [inline install/kubernetes/cilium/values.yaml]: Nitpick: the actual option name differs

```suggestion
  # -- Enforce https for host having matching TLS host in Ingress.
  # Incoming traffic to http listener will return 308 http error code with respective location in header.
  enforceHttps: true
```

**Review by qmonnet — APPROVED** [review]
Not super familiar with all the parts involved, so I've glanced rather quickly over some of the changes. Looks good overall, thank you!

File: install/kubernetes/cilium/values.yaml
Comment (sayboras) [reply install/kubernetes/cilium/values.yaml]: thanks, I wait for current runing CI jobs to complete and address the above.

**Comment by sayboras** [comment]
Runtime failed due to below error, which is highlighted in master as well.

```
pkg/maps/lbmap/maglev_test.go:55:3: unknown field 'MaxEntries' in struct literal of type InitParams
```

File: operator/pkg/ingress/envoy_config.go
Comment (jrajahalme) [reply operator/pkg/ingress/envoy_config.go]: No need to validate character set if it is already validated.

**Review by jrajahalme — APPROVED** [review]
Assuming that the Ingress spec is validated not to contain `*`, `+`, `?`, or any other regex control characters other than dots and a possible `*.` combo in the beginning.

If Ingress spec validation is laxer than that then our regex conversion would need additional validation.

**Comment by sayboras** [comment]
/test-1.23-net-next

**Review by joestringer — APPROVED** [review]
Helm changes LGTM. We're also missing kubernetes codeowner review, but the only directly-relevant thing I see is the `install/kubernetes/...` helm bits, so I think we're good on that front.

Can we validate Jarno's assumption above? With that confirmed, I think this should be good to :ship:

**Comment by sayboras** [comment]
thanks for your comments, please find below my verification, basically only [a-z0-9] and dot are allowed.

```
# doesn't start with * (so *, +, ? characters are not allowed) 
a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')

# start with *. (so * in the middle is not allow)
a wildcard DNS-1123 subdomain must start with '*.', followed by a valid DNS subdomain, which must consist of lower case alphanumeric characters, '-' or '.' and end with an alphanumeric character (e.g. '*.example.com', regex used for validation is '\*\.[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')
```

**Pull request merged**
